File: test/app_search_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:installed_apps/app_info.dart';
import 'package:FuseLauncher/app_search.dart';

AppInfo app(String name) => AppInfo(
      name: name,
      icon: null,
      packageName: name.toLowerCase().replaceAll(' ', '.'),
      versionName: '1.0.0',
      versionCode: 1,
      builtWith: BuiltWith.flutter,
      installedTimestamp: 0,
    );

void main() {
  test('ranks exact, prefix and word-start matches above substrings', () {
    final apps = ['ibisPaint X', 'KernelSU Next', 'myXL', 'Sandbox', 'Termux', 'X']
        .map(app)
        .toList();

    expect(searchApps(apps, 'x').map((a) => a.name).toList(), [
      'X',
      'ibisPaint X',
      'KernelSU Next',
      'myXL',
      'Sandbox',
      'Termux',
    ]);
  });

  test('name prefix beats word start, and non-matches drop out', () {
    final apps = ['ibisPaint X', 'Xodo', 'Termux', 'Clock'].map(app).toList();

    expect(searchApps(apps, 'x').map((a) => a.name).toList(),
        ['Xodo', 'ibisPaint X', 'Termux']);
  });

  test('empty query returns the list untouched', () {
    final apps = ['B', 'A'].map(app).toList();
    expect(searchApps(apps, ''), same(apps));
  });
}

File: app/src/test/kotlin/com/kayfahaarukku/fuselauncher/apps/AppSearchTest.kt
package com.kayfahaarukku.fuselauncher.apps

import org.junit.Assert.assertEquals
import org.junit.Assert.assertSame
import org.junit.Test

private fun app(name: String) = LauncherApp(
    packageName = name.lowercase().replace(' ', '.'),
    componentName = "${name.lowercase().replace(' ', '.')}/.Main",
    label = name,
)

class AppSearchTest {

    @Test
    fun `ranks exact, prefix and word-start matches above substrings`() {
        val apps = listOf("ibisPaint X", "KernelSU Next", "myXL", "Sandbox", "Termux", "X")
            .map(::app)

        assertEquals(
            listOf("X", "ibisPaint X", "KernelSU Next", "myXL", "Sandbox", "Termux"),
            searchApps(apps, "x").map { it.label },
        )
    }

    @Test
    fun `name prefix beats word start, and non-matches drop out`() {
        val apps = listOf("ibisPaint X", "Xodo", "Termux", "Clock").map(::app)

        assertEquals(
            listOf("Xodo", "ibisPaint X", "Termux"),
            searchApps(apps, "x").map { it.label },
        )
    }

    @Test
    fun `empty query returns the list untouched`() {
        val apps = listOf("B", "A").map(::app)
        assertSame(apps, searchApps(apps, ""))
    }
}
